accepted/0000-retail-pxe-database.md

```diff
@@ -0,0 +1,259 @@
+
+Column `mac_address` specify MAC address of the terminal. Similar column is also present in `rhnServerNetInterface` table, however that information is populated only after hardware refresh is scheduled, which is done only after successful deployment. Whereas pxe entries needs to be available before full deployment because of cases when kernel version of inird is different to kernel version in image and terminal requires reboot to correct kernel version.
+
+Another option is that event handler creates the correct entry in `rhnServerNetInterface` table.
+
+### Detecting pxeserver_id
+
+In scenarios where terminals are connected to the branch server proxy, `pxeserver_id` is currently assumed to be branch server proxy and can be looked up in database by event handler. However architectures where proxy server is not involved such detection will not work (for example Hub scenarios where peripheral server is used instead of proxy server).
+
+We propose to solve this by **adapting saltboot to send FQDN of `next-server`**, that means server which provided initrd and kernel during PXE boot. This can be obtained by parsing DHCP response for `next-server` field. DHCP response is stored and available in dracut environment where saltboot is operating. Mandatory requirement in this case would be that **reverse lookup of IP address must be correctly configured** as `next-server` fields is IP address.
+
+Alternative is to use DNS service to lookup `tftp` CNAME, this is discussed in alternatives below.
+
+When the FQDN of `next-server` is not in the Uyuni database, such is the case when terminal is connected directly to the Uyuni, then `pxeserver_id` would be set to `null`.
+
+This is related to the [Hub Retail extension](https://github.com/uyuni-project/uyuni-rfc/pull/45) discussion about how should Retail architecture look like in Hub environment.
+
+## Storing data in database
+
+As data are provided within salt event, implementing new event handler `SaltbootPxeUpdateEvent` with accompanying `SaltbootPxeUpdateMessageAction` should be sufficient.
+
+Event handler should also schedule generation of new PXE entry.
+
+## Exposing PXE entries to users
+
+In current implementation, consumer of the `pxe_update` event data is `pxe/terminal_entry` state which is part of PXE formula.
+
+In case of keeping the final action in salt space, there is a need to provide pillar data to the state:
+
+  * by supplying directly to state when scheduling pxe generating action.
+  * by exposing database data to salt pillar system.
+
+Exposing PXE data is not hard requirement but may have some benefit like enabling user to manually apply `pxe/terminal_entry` or similar state to (re)generate pxe entries. This option can save resources by making UI/XML-RPC API changes non-essential.
+
+Database can be exposed to pillar system by external pillar mechanism connecting to the database using postgresql external pillar:
```

Comment:
I think we will need a solution to avoid file pillars at some point, but should be addressed in a separate PR.

I do not see a compelling reason to introduce a postgres external pillar only for the use case under discussion - rather, all generated pillars should be moved off the filesystem and onto the database in one shot, as part of an ad-hoc effort.

So for now we are only left with either generating file-pased pillars or exposing an XMLRPC endpoint.

Reply:
Please then move this section to Alternatives, reporting the reason why it's not tackled this way for now.